DAOS-14539 client: intercept execve() and it variants

wiliamhuang opened the pull request:
Features: pil4dfs

envp different from environ could be passed into execve() and execvpe() to override the default environ. Under rare circumstances (e.g., scons, git), IL related envs are not propagated from parent process to newly created child processes. D_IL_ENFORCE_EXEC_ENV=1 will force libpil4dfs related env variables appended when exec() is called. This is turned off as default.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/3/execution/node/1162/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

wiliamhuang commented:
Just converted this PR back to draft. I will spend a little time to double check all resources are cleaned up before exec().

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/7/execution/node/1319/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/7/execution/node/1411/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/7/execution/node/1365/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/9/execution/node/1318/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13519/10/testReport/

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/11/execution/node/302/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/11/execution/node/346/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/11/execution/node/402/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/11/execution/node/301/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/11/execution/node/280/log

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13519/14/testReport/

mchaarawi commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+					pil4dfs_in_preload = true;
+			} else if (strncmp(envp[num_entry], "D_IL_REPORT", 11) == 0) {
+				report_included = true;
+			} else if (strncmp(envp[num_entry], "DAOS_MOUNT_POINT", 16) == 0) {
+				mp_included = true;
+			} else if (strncmp(envp[num_entry], "DAOS_POOL", 9) == 0) {
+				pool_included = true;
+			} else if (strncmp(envp[num_entry], "DAOS_CONTAINER", 14) == 0) {
+				cont_included = true;
```
what about other env variables that DAOS users might set? 
D_LOG_MASK, D_LOG_FILE, etc ?

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+extern char **__environ;
+
+/* This number may be updated later to be consistent!*/
+#define N_ENV_CHECK (7)
```
Consistent with what?

wiliamhuang replied:
The comment was revised. Thank you!

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
  */
 
+static char **
+pre_envp(char *const envp[])
```
Perhaps you could also explain why this is needed as well?  Environment variables are usually inherited by default by both child processes and their children, why is there a hook to re-add them needed?

wiliamhuang replied:
Comments are added. Yes. env variables are normally inherited from the parent process. envp could be passed into execve() and execvpe() to override the default envp. env was cleared in many child processes created by python. Intercepting execve() etc. can force the needed env set if absent.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+		num_entry = 0;
+	} else {
+		while (envp[num_entry]) {
+			if (strncmp(envp[num_entry], "LD_PRELOAD", 10) == 0) {
```
This is a very error prone way of handling strings although it may be that the numbers you have in this PR are correct, please consider something like the following or better still wrapping this up in a macro so the string can only be passed in once.
```suggestion
			if (strncmp(envp[num_entry], "LD_PRELOAD", sizeof"LD_PRELOAD")) == 0) {
```

wiliamhuang replied:
Updated as you suggested. Thank you!
I will replace them with a macro as you suggested when I refresh my PR next time.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+		}
+	}
+
+	/* The fd for daos logging may be not valid here, so use plain printf */
```
This comment applies to many places in this function, assuming it's correct then put it once in the function description at the start.

wiliamhuang replied:
Moved this line after function description. Thank you!

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+
+	pil4df_path = query_pil4dfs_path();
+	len2        = strnlen(pil4df_path, PATH_MAX);
+	/* copy existing entries */
```
You need a better comment for what this loop is doing.

wiliamhuang replied:
Thank you! I rearranged the code here and added comments.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+		if (fs_root) {
+			rc = asprintf(&str_mp, "DAOS_MOUNT_POINT=%s", fs_root);
+			if (rc < 0) {
+				printf("Error: failed to allocate memory for DAOS_MOUNT_POINT "
+				       "env!\n");
```
Applies elsewhere but please don't use ! in error messages!

It looks like this would all fit on line line as well.

wiliamhuang replied:
Thank you! Removed "!". What did you mean "all fit on line line"? Is it "all fit on one line"?

ashleypittman replied:
You could remove the newline so the code could fit on one line of the source file rather than being wrapped onto 2.  We allow 100 characters per line.

wiliamhuang replied:
Thank you! I double-checked this line. There are still 101 characters in this line after "\n" is removed.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+	return (char **)envp;
+}
+
+/* check whether fd 0, 1, and 2 are located on DFS mount or not. If yes, reopen files and
+ * set offset.
```
Is offset meaningful here?  At least for 1 and 2 it could be, but if it is then this seems like something that would benefit from a test.

wiliamhuang replied:
One test extracted from "configure" is added in dfuse/bash.py. The code here was removed and put in dup2() as a special case for bash/sh.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+			open_flag = file_list[idx]->open_flag;
+
+			/* get a real fd from kernel */
+			fd_tmp = libc_open(file_list[idx]->path, open_flag);
```
Do you have any consideration for if the path has changed, this could legitimately return a different file or ENOENT I assume?

wiliamhuang replied:
The code is moved to dup2(). If libc_open() fails, then dup2() fails with libc_open()'s errno.

ashleypittman commented on src/tests/ftest/daos_test/dfuse.py:
```diff
@@ -91,8 +91,15 @@ def run_test(self, il_lib=None):
             daos_test_env['D_LOG_MASK'] = 'INFO,IL=DEBUG'
-
-        command = [self.daos_test, '--test-dir', mount_dir, '--io', '--stream']
+            if il_lib == 'libpil4dfs.so':
+                daos_test_env['D_IL_REPORT'] = '0'
+                daos_test_env['DAOS_MOUNT_POINT'] = f'{mount_dir}'
+                daos_test_env['DAOS_POOL'] = f'{self.pool.identifier}'
+                daos_test_env['DAOS_CONTAINER'] = f'{self.container.uuid}'
+                daos_test_env['D_IL_MAX_EQ'] = '2'
+                daos_test_env['D_IL_ENFORCE_EXEC_ENV'] = '1'
```
This is not a new test, what has changed to mean these are required now?

wiliamhuang replied:
The new tests in dfuse_test will check libpil4dfs related env to make sure they are enforced to set in child process.

ashleypittman commented on src/tests/ftest/dfuse/bash.py:
```diff
@@ -119,6 +119,9 @@ def run_bashcmd(self, il_lib=None, compatible_mode=False):
                             f'realpath {link_name}',
                             f'head {fuse_root_dir}/src.c',
                             f'tail {fuse_root_dir}/src.c',
+                            'printf "exec 5>config.log\n{\ncat <<EOF\n------\nEOF\n} >&5\n"'
+                            f'> {fuse_root_dir}/1.sh; cd {fuse_root_dir}; '
+                            f'/usr/bin/sh {fuse_root_dir}/1.sh',
```
I assume this is covered in https://github.com/daos-stack/daos/pull/13681 in which case please drop the changes to this file.

wiliamhuang replied:
ok. I will drop the changes here. Thank you!

ashleypittman reviewed (commented):
Neither the ticket or PR description elaborates, why do we need to intercept evecve?  Is this so that we can run binaries from daos using pil4dfs without a dfuse backend?

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/16/execution/node/398/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/16/execution/node/358/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/16/execution/node/302/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/16/execution/node/374/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/16/execution/node/393/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/18/execution/node/382/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/18/execution/node/377/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/18/execution/node/303/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/18/execution/node/406/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/18/execution/node/415/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/19/execution/node/416/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/19/execution/node/370/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/19/execution/node/334/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/19/execution/node/383/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/19/execution/node/335/log

wiliamhuang commented on src/client/dfuse/pil4dfs/hook.c:
```diff
@@ -308,6 +309,24 @@ determine_lib_path(void)
 	}	
 	D_FREE(lib_dir_str);
 
+	pos = strstr(read_buff_map, "libpil4dfs.so");
```
The comment was revised.

wiliamhuang commented:
I will set D_IL_ENFORCE_EXEC_ENV off as default when I refresh this PR.

daltonbohning commented on src/tests/ftest/daos_test/dfuse.py:
```diff
@@ -91,8 +91,15 @@ def run_test(self, il_lib=None):
-        command = [self.daos_test, '--test-dir', mount_dir, '--io', '--stream']
+            if il_lib == 'libpil4dfs.so':
+                daos_test_env['D_IL_REPORT'] = '0'
+                daos_test_env['DAOS_MOUNT_POINT'] = f'{mount_dir}'
+                daos_test_env['DAOS_POOL'] = f'{self.pool.identifier}'
+                daos_test_env['DAOS_CONTAINER'] = f'{self.container.uuid}'
```
No need to use f-strings for these. Also, any reason why the container label wouldn't work?
```suggestion
                daos_test_env['DAOS_MOUNT_POINT'] = mount_dir
                daos_test_env['DAOS_POOL'] = self.pool.identifier
                daos_test_env['DAOS_CONTAINER'] = self.container.identifier
```

wiliamhuang replied:
Container label should work. I will fix them. Thank you very much!

wiliamhuang replied:
Fixed as you suggested. Thank you!

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13519/20/testReport/

wiliamhuang commented:
One ftest failed. It will need a bit of work to fix the issue.

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13519/21/display/redirect

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13519/21/display/redirect

daosbuild1 commented:
Test stage Unit Test bdev on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13519/21/display/redirect

daosbuild1 commented:
Test stage Unit Test on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13519/21/display/redirect

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13519/22/display/redirect

wiliamhuang commented:
@ashleypittman @mchaarawi Could you please review this PR? Thank you very much!

ashleypittman commented on src/tests/suite/dfuse_test.c:
```diff
@@ -48,6 +49,9 @@ print_usage()
 	print_message("dfuse_test -l|--lowfd\n");
+	print_message("dfuse_test -e|--exec\n");
+	/* verifyenv is only run by exec test. Should not be executed directly */
+	/* print_message("dfuse_test -v|--verifyenv\n");                       */
```
Is this is internal then only support the long option here.  -v could well use run by someone wanting to check the version.
```suggestion
	/* print_message("dfuse_test    --verifyenv\n");                       */
```

wiliamhuang replied:
Thank you! I will remove "-v".

ashleypittman commented on src/tests/suite/dfuse_test.c:
```diff
@@ -654,9 +787,11 @@ main(int argc, char **argv)
 					       {NULL, 0, NULL, 0}};
 
-	while ((opt = getopt_long(argc, argv, "aM:imsdl", long_options, &index)) != -1) {
+	while ((opt = getopt_long(argc, argv, "aM:imsdlev", long_options, &index)) != -1) {
```
You can support the long option only by leaving the short option out of this string.
```suggestion
	while ((opt = getopt_long(argc, argv, "aM:imsdle", long_options, &index)) != -1) {
```

wiliamhuang replied:
Thank you! I will fix it as you suggested.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,347 @@ new_readdir(DIR *dirp)
+extern char **__environ;
+
+/* This is the number of environmental variables that would be forced to set in child process. */
+#define N_ENV_APPEND (7)
```
Should this be forwarding n D_LOG_FILE as well?

wiliamhuang replied:
Thank you! I will add "D_LOG_FILE". Shall we add "DD_MASK", "DD_SUBSYS", and "D_LOG_MASK" too?

ashleypittman replied:
Yes, I think so.

It would be a good idea to re-write this code to use and array (possilby you had it like that before?) so that adding new ones is easy as I'm sure we're going to find new entries to add to the list at some point.  I think LD_PRELOAD is the only one that needs a special case here, the rest all copy from the current process I assume?

wiliamhuang replied:
Thank you! I will look into rewriting the code here as you suggested.

wiliamhuang replied:
I just updated the code to loop over a given env list to append. Thank you!

wiliamhuang commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -5680,6 +5983,13 @@ init_myhook(void)
 		d_freeenv_str(&env_log);
 	}
+	enforce_exec_env = false;
+	d_agetenv_str(&env_exec, "D_IL_ENFORCE_EXEC_ENV");
```
Thank you! I will use d_getenv_bool() instead.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,282 @@ new_readdir(DIR *dirp)
+	for (j = 0; j < LEN_ENV_LIST; j++) {
+		/* env is not found. Need to be appended if present in current process. */
+		if (!env_found[j]) {
+			rc = d_agetenv_str(&env_value, env_list[j]);
```
Does this work?  At this point if the environment has been cleared then this might fail?  In which case you'd need to sample the value in init() and keep it around to use it here.

wiliamhuang replied:
I think this should work. unsetenv() is not called in the case of scons, so env variables should be there.

ashleypittman replied:
It seems likely that this will work for programs that call evecve without the env set but not for functions that call unsetenv and then execv, this is something we should test for and if necessary note.

wiliamhuang replied:
I believe applications call unsetenv() to clear only common system env or the env they set. I only observed "MAKEFLAGS" cleared by unsetenv() two times during daos building with scons.
We can intercept unsetenv(env) and ignore the operation if env is in our list and D_IL_ENFORCE_EXEC_ENV == 1.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,282 @@ new_readdir(DIR *dirp)
+				       env_list[j]);
+				goto err_out2;
+			}
+			rc = asprintf(&env_buf_list[j], "%s=%s", env_list[j], env_value);
```
This looks incorrect for the -DER_NONEXIST case.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,282 @@ new_readdir(DIR *dirp)
+/* This is the number of environmental variables that would be forced to set in child process.
+ * "LD_PRELOAD" is a special case and it is not included in the list.
+ */
+#define LEN_ENV_LIST (10)
```
You could now use ARRAY_SIZE(env_list) for this so to avoid having to update it.  You'd need to make env_list a static global or this a local variable though.

wiliamhuang replied:
Thank you! Fixed as you suggested.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,282 @@ new_readdir(DIR *dirp)
+	bool   pil4dfs_in_preload         = false;
+
+	/* "LD_PRELOAD" is a special case which needs .*/
+	char   env_list[LEN_ENV_LIST][32] = {"D_IL_REPORT", "DAOS_MOUNT_POINT", "DAOS_POOL",
```
Is the 32 or LEN_ENV_LIST required here?  As you're populating this list you can probably let the compiler do the work here.  See https://github.com/daos-stack/daos/blob/master/src/cart/crt_init.c#L21 for now to do this dynamically without using constants.
```suggestion
	char   *env_list[] = {"D_IL_REPORT", "DAOS_MOUNT_POINT", "DAOS_POOL",
```

wiliamhuang replied:
Fixed as you suggested. Thank you very much!

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/26/execution/node/1320/log

wiliamhuang commented:
The failures in Functional Hardware Medium Verbs Provider are not related.
@ashleypittman @mchaarawi Could you please review?

daltonbohning reviewed (commented):
ftest code changes LGTM

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,284 @@ new_readdir(DIR *dirp)
- *		return pid;
- *	}
- *}
+extern char **__environ;
```
If I've read the man page correctly then this is also available as environ and available through a standard header file.

wiliamhuang replied:
You are right.
"/usr/include/unistd.h" does have "extern char **__environ;" and "char **environ". I will fix this. Thank you!

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,284 @@ new_readdir(DIR *dirp)
+ * env when needed.
+ * pre_envp() scans provided envp to look for libpil4dfs related envs (env_list[]) and returns a
+ * new envp with missing libpil4dfs related envs appended.
+ * The fd for daos logging may be not valid here, so printf() is used in this function.
```
This is a callback between for and exec, I'm not sure printf() would always work here but for errors then fprintf() to stderr would be better.

wiliamhuang replied:
ok. I will use fprintf(stderr, ). Thank you!

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,284 @@ new_readdir(DIR *dirp)
+			/* The list of env is not too long. We use a simple loop to lookup for
+			 * simplicity. This function is not performance critical.
+			 */
+			for (i = 0; i < ARRAY_SIZE(env_list); i++) {
```
There's a case here where this code will be invoked but not required to.  If LD_PRELOAD is being used but one of the env vars you're testing for is not set then this will be detected and the code will start to modify the env, but if the env isn't set then it won't be able to be copied so in effect this will make a copy but then not modify it.

To catch this case then I think this while code should loop and call d_agetenv_str() first for each entry and then only search the list for ones which are set in the parent.

wiliamhuang replied:
Thank you very much! Revised as you suggested.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,284 @@ new_readdir(DIR *dirp)
+		next_execve = dlsym(RTLD_NEXT, "execve");
+		D_ASSERT(next_execve != NULL);
+	}
+	return next_execve(filename, argv, pre_envp(__environ));
```
I still think this is unlikely to work, if a env var has been removed then d_agetenv_str() on it would fail.

wiliamhuang replied:
ok. I will add the interception of unsetenv() then. Thank you!

ashleypittman replied:
It would be simpler and more reliable to save the env vars that might be required at the time that the parent process itself is initialized.

wiliamhuang replied:
Saving the value at initialization might cause inconsistency in application like bash. If "export" is used to update env in bash script, the saved value would be out of dated. Using env does not have such issue.

ashleypittman replied:
That's true, but if that's the case then it would be set and the saved value wouldn't be needed.

getenv() probably just checks the environ list so having code which checks environ and adds anything that's missing from environ isn't going to achieve much.

wiliamhuang replied:
It would be more reasonable to let the lib/application set default value if an env is not set instead of using a saved value at the initialization phase.

Assuming env_var's default value is 0.
bash runs a bash script with env_var=1.
"unset env_var" is invoked inside bash script to clear env on purpose.
"env_var=1" will still be used in following new processes.

Relying on getenv would be more consistent to me.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -5673,13 +5913,19 @@ init_myhook(void)
 		if (strncmp(env_log, "0", 2) == 0 || strncasecmp(env_log, "false", 6) == 0)
 			report = false;
 		d_freeenv_str(&env_log);
+	} else if (rc == -DER_NONEXIST) {
+		d_setenv("D_IL_REPORT", "0", 1);
```
I don't think you should be setting these.  I assume the intention is to save them for if required later in exec but that has two problems:

1. They might have been stripped from the environment
2. When running pil4dfs then the child and grandchild will have different environment

Assuming the intention here is to save them for later then use a static global variable for this.

wiliamhuang replied:
same as the above comment. 

It would lead to better consistency to let the lib/application set default value if an env is not set instead of using a saved value at the initialization phase.

Assuming env_var's default value is 0.
bash runs a bash script with env_var=1.
"unset env_var" is invoked inside bash script to clear env on purpose.
"env_var=1" will still be used in following new processes.

For the case of creating a child process with stripped env, parent process does not unset env in environ. Instead, the parent process just passes a new env list in exec().

ashleypittman replied:
The above discussion was in the context of a PR where execv was being modified but now the PR only modified execve.

I still don't see any case for setting this here.  If a user has set it and it's propagated then fine, but why would a process turn it on for it's child processes?

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,284 @@ new_readdir(DIR *dirp)
+			    "D_IL_MAX_EQ", "D_LOG_FILE", "D_IL_ENFORCE_EXEC_ENV",  "DD_MASK",
+			    "DD_SUBSYS", "D_LOG_MASK"};
+
+/* Environmental variables could be cleared in some applications. To make sure all libpil4dfs
```
I'm trying to understand the error handling here, it seems this function is unlikely to fail, however if it does then it would do it in a way that it does nothing which were it to cause problems would be very difficult to debug so this should probably return an error and then propagate to the user.  I would probably just remove the prints about memory allocation failure as well.

wiliamhuang replied:
Thank you! Revised as you suggested.

ashleypittman commented:
FYI - this has a conflict now with some other changes I just made to the dfuse_test.

wiliamhuang commented:
> FYI - this has a conflict now with some other changes I just made to the dfuse_test.

Thank you for letting me know! I will fix them when I refresh my PR.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/30/execution/node/1430/log

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/31/execution/node/1199/log

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/32/execution/node/458/log

wiliamhuang commented:
@ashleypittman The failures in CI are not relevant. Could you please review and provide your feedback?
We only append env if the provided env list is different from environ. I think relying on getenv would achieve better consistency especially for bash.
When a parent process creates a child process with a stripped env, it uses an env list different from environ. If a process updates env, we should honor such changes.

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/33/execution/node/1198/log

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/34/execution/node/458/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/35/execution/node/333/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/35/execution/node/380/log

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,350 @@ new_readdir(DIR *dirp)
+	if (envp == environ)
+		return (char **)envp;
+
+	/* check D_IL_ENFORCE_EXEC_ENV env in case it was updated. */
+	rc = d_getenv_bool("D_IL_ENFORCE_EXEC_ENV", &enforce_exec_env);
+	if (rc == -DER_NONEXIST || enforce_exec_env == false)
+		return (char **)envp;
```
If this environment has been cleaned then this won't be set so the code would never be invoked, at least for the exec case.

You also don't want to read the same environment variable more than once so can you read this at the time daos_init() is called or elsewhere where the library is setup and save the result for use here.

wiliamhuang replied:
Flag enforce_exec_env is set in the constructor function by getenv. We do not monitor whether this env is updated or not. In the case of running a bash script, env "D_IL_ENFORCE_EXEC_ENV" could be updated inside the script. Checking it again will make sure we do get an updated value.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,350 @@ new_readdir(DIR *dirp)
+	num_env_append = 0;
+
+	/* check whether "LD_PRELOAD" exists in environ */
+	rc = d_agetenv_str(&env_value, "LD_PRELOAD");
+	if (rc == -DER_NONEXIST) {
+		return (char **)envp;
```
So if `LD_PRELOAD` isn't set then this code does not set it?  Isn't adding this if it's missing the point of this PR?

wiliamhuang replied:
If "LD_PRELOAD" is unset inside a bash script, we should honor it.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,350 @@ new_readdir(DIR *dirp)
+		pil4dfs_set_preload = true;
+		num_env_append++;
+	}
+	free(env_value);
```
I'm not completely up to date here but should this use `d_freeenv_str`?

wiliamhuang replied:
You are right. I will replace "free" with "d_freeenv_str". Thank you!

wiliamhuang replied:
Fixed as you suggested. Thank you!

wiliamhuang commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -2011,6 +2027,14 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
 	file_list[idx_fd]->open_flag   = oflags;
 	/* NEED to set at the end of file if O_APPEND!!!!!!!! */
 	file_list[idx_fd]->offset = 0;
+	if (strncmp(full_path, "/", 2) == 0)
```
Yes. I though about using "STR_AND_SIZE" previously and chose not changing it. This string is short and current form is easier to read.

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13519/37/display/redirect

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/38/execution/node/270/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/38/execution/node/292/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/38/execution/node/369/log

ashleypittman commented:
@wiliamhuang I had missed the changes in https://github.com/daos-stack/daos/pull/13519/commits/ebf7a5b944877bdcd141cdfc50cd954e408a4670 so that this PR no longer modifies `execv` but only `execve` variants.

The assumption now appears to be that the process still has everything set in the environment but that if it's using `execve` then this should add back in things from the environment.

I'll re-review with this in mind.

wiliamhuang commented:
> @wiliamhuang I had missed the changes in [ebf7a5b](https://github.com/daos-stack/daos/commit/ebf7a5b944877bdcd141cdfc50cd954e408a4670) so that this PR no longer modifies `execv` but only `execve` variants.
> 
> The assumption now appears to be that the process still has everything set in the environment but that if it's using `execve` then this should add back in things from the environment.
> 
> I'll re-review with this in mind.

Right. We only append env for execve, execvpe, and fexecve where an env list different from environ can be passed in. If envs in environ updated (modified/removed/add), we do nothing since we assume that are done on purpose.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/40/execution/node/1430/log

wiliamhuang commented:
The failures in hardware medium verb providers are not relevant. I can rerun the hardware tests.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/42/execution/node/1451/log

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,349 @@ new_readdir(DIR *dirp)
+			rc = ENOMEM;
+			goto err_out1;
+		}
+		(*new_envp)[idx_preload] = new_preload_str;
```
Do you need to free the old string?

wiliamhuang replied:
I prefer not freeing the old string. The string could be static.

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,349 @@ new_readdir(DIR *dirp)
+		return 0;
+
+	/* the new envp holds the existing envs & the envs forced to append plus NULL at the end */
+	*new_envp = malloc(sizeof(char *) * (num_entry + num_env_append + 1));
```
calloc would be better here as it would be zeroed memory.

wiliamhuang replied:
Revised as you suggested. Thank you!

ashleypittman commented on src/client/dfuse/pil4dfs/int_dfs.c:
```diff
@@ -3310,70 +3335,349 @@ new_readdir(DIR *dirp)
+	}
+
+	/* append NULL pointer to the end of new_envp */
+	(*new_envp)[i] = NULL;
```
using calloc above would avoid the need for this.

wiliamhuang replied:
Revised as you suggested. Thank you!

wiliamhuang commented:
@ashleypittman I removed unneeded setenv() for "D_IL_REPORT" and "D_IL_ENFORCE_EXEC_ENV". The default value will be set in case env is not found. Thank you!

ashleypittman commented:
> @ashleypittman I removed unneeded setenv() for "D_IL_REPORT" and "D_IL_ENFORCE_EXEC_ENV". The default value will be set in case env is not found. Thank you!

I think the same applies to `D_IL_MAX_EQ` as well.

wiliamhuang commented:
> > @ashleypittman I removed unneeded setenv() for "D_IL_REPORT" and "D_IL_ENFORCE_EXEC_ENV". The default value will be set in case env is not found. Thank you!
> 
> I think the same applies to `D_IL_MAX_EQ` as well.

Thank you! Just fixed it.

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/45/execution/node/1198/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13519/46/testReport/

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/46/execution/node/1539/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13519/47/execution/node/468/log

wiliamhuang commented:
The failures of "Functional Hardware Medium Verbs Provider " are not relevant to libpil4dfs.

mchaarawi commented:
> The failures of "Functional Hardware Medium Verbs Provider " are not relevant to libpil4dfs.

what is the ticket number associated with the HW medium failure?
